### PR TITLE
osd: fix reference count, rare race condition etc.

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -3050,6 +3050,11 @@ PG *OSD::get_pg_or_queue_for_pg(const spg_t& pgid, OpRequestRef& op)
   PG *out = NULL;
   if (wlistiter == session->waiting_for_pg.end()) {
     out = i->second;
+    // On succeeding in getting a PG, put a reference on it before
+    // we drop the pg_map_lock and give it back to caller.
+    // This is necessary because we may race with _remove_pg(),
+    // who will remove the last reference and destruct PG in most case.
+    out->get("get_pg_or_queue_for_pg");
```
Comment (tchaikov) [inline src/osd/OSD.cc]: but they are both guarded by osd_lock, i believe?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -6901,6 +6901,10 @@ void OSD::_committed_osd_maps(epoch_t first, epoch_t last, MOSDMap *m)
         epoch_t bind_epoch = osdmap->get_epoch();
         service.set_epochs(NULL,&up_epoch, &bind_epoch);
 	do_restart = true;
+        int r;
+	set<int> avoid_ports;
+
+        start_waiting_for_healthy();
```
Comment (tchaikov) [inline src/osd/OSD.cc]: i am not sure we need to optimise for rebind error at the expense of introducing more entropy into the code.

**Comment by tchaikov** [comment]
apart from 9f2a1c9 and 8f54967 , the commits look good to me.

**Comment by xiexingguo** [comment]
@tchaikov  Dropped those two changes. Thanks for the review.

**Comment by tchaikov** [comment]
lgtm

**Comment by tchaikov** [comment]
lgtm.

**Pull request merged**